ui/src/app/utils/cluster-initializer.spec.tsx

```diff
@@ -0,0 +1,305 @@
+      throw new Response(null, {status: 404});
+    });
+    mockCreateCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      return {...baseCluster, status: ClusterStatus.Creating};
+    });
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Starting},
+      {...baseCluster, status: ClusterStatus.Running}
+    ]);
+    const initializer = createInitializer();
+    const cluster = await runInitializerAndTimers(initializer);
+
+    expect(mockCreateCluster).toHaveBeenCalled();
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('should delete cluster if in an error state', async() => {
+    // A cluster in an error state should trigger a deletion request.
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Creating},
+      {...baseCluster, status: ClusterStatus.Error},
+    ]);
+    mockDeleteCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      return {};
+    });
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Deleting},
+      {...baseCluster, status: ClusterStatus.Deleting},
+    ]);
+    // After some period of "deleting" status, we expect the cluster to become nonexistent...
+    mockGetCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      throw new Response(null, {status: 404});
+    });
+    // which should trigger a creation request...
+    mockCreateCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      return {...baseCluster, status: ClusterStatus.Creating};
+    });
+    // and eventually give us a good cluster.
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Starting},
+      {...baseCluster, status: ClusterStatus.Running}
+    ]);
+
+    const initializer = createInitializer();
+    const cluster = await runInitializerAndTimers(initializer);
+
+    expect(mockDeleteCluster).toHaveBeenCalled();
+    expect(mockCreateCluster).toHaveBeenCalled();
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('Should recover from intermittent 500s', async() => {
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Creating},
+      {...baseCluster, status: ClusterStatus.Creating},
+    ]);
+    mockGetCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      throw new Response(null, {status: 503});
+    });
+    mockGetCluster.mockImplementationOnce(async(workspaceNamespace) => {
+      throw new Response(null, {status: 503});
+    });
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Running},
+    ]);
+
+    const initializer = createInitializer();
+    const cluster = await runInitializerAndTimers(initializer);
+
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('Should give up after too many server errors', async() => {
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Creating},
+    ]);
+    for (let i = 0; i < 10; i++) {
```

Comment:
nit: I looked for a `mockImplementationNTimes()` but didn't see anything. It might not hurt to write your own if you need to do this in more than one place.